### PR TITLE
Offline Pages : PageList Item Progress State 

### LINKED ISSUE #11127: Show "uploading" progress indicator on PageList items
We show a progress indicator when a post is being uploaded, but there is no visual cue when a page is being uploaded. 

Goal of this task is to 
- [ ] Add a progress bar 

to page list items. 

The design can be copied from PostList.

### PR DESCRIPTION
**Pull request by jd-alexander**
Fixes #11127

## Findings
The Post List support showing progress for uploads and since pages are posts then the implementation for the Post List can be adapted for the Page List. 

## Solution
To utilize existing classes and logic that's used for the Post List progress state to power the Page List. This primarily involves creating new classes for this behavior since refactoring the logic for the Post List to support Pages might not be in our best interest at this point since we aren't sure what will need to be shared. 

An additional improvement that was made in this PR was to create a component similar to `PostListEventListener` called `PageListEventListener` so that the calls and events from the `dispatcher` are encapsulated in a central component to reduce the clutter in the `PagesViewModel`. While doing the review, I would appreciate if the coroutine and overall thread usage could be checked to ensure it is correct. 

## Testing

For all these tests it's best to change the device's network type to GSM and it's signal strength to poor to observe the state changes in the list. 

### Upload Page Offline 

1. Turn on airplane mode. 
2. Create a new post. 
3. Click publish.
4. Ensure the post you published is visible and turn airplane mode off.
5. Observe that the state changes cause the `Page List Item` to display an indeterminate `ProgressBar` and an overlay on the row contents. 

Before | After 
--------|-------
        |    

### Upload Page with Image/Video Offline 

1. Turn on airplane mode. 
2. Create a new post and add an image or video. 
3. Click publish.
4. Ensure the post you published is visible and turn airplane mode off.
5. Observe that the state changes cause the `Page List Item` to display a determinate `ProgressBar` and an overlay on the row contents. 
6. The `ProgressBar` should show the percentage of the current upload. 
7. Also, observe that when the media upload is complete the `ProgressBar` switches back to `Indeterminate` to finish the upload. 

Before | After 
--------|-------
        |    

### Other testing considerations

1. Ensure that both cases above work when the device is online and didn't go into airplane mode. 
2. While the upload is taking place, navigate back to the `My Site` page and then click the `Pages` button and ensure the correct state is being shown. 
3. Also, ensure that the progress state is always `Hidden` once the upload is complete or an error takes place. (Error wasn't implemented in this PR, it possibly just switches back to the local changes label and shows the error in the notification section.)
4. There are several events within `PostListEventListener` that inspired the `PageListEventListener`It's important that a comparison between both behaviors are done to ensure that the upload state is reacting effectively to this. Some of the events that weren't copied over were related to showing error and success states so those were left for future PRs. Only events that modify the upload state were utilized. 
5. The layout wasn't modified to support the `status` label being above the overlay since it was outside the scope of this PR, so in the tests that will be noticeable. 
6. When testing with video, the video optimization can be enabled in the App Settings so that the progress from that can also be propagated and verified. 

## Reviewing 
Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] I have considered adding accessibility improvements for my changes.
- [x] If it's feasible, I have added unit tests.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt [hunk 1]
```diff
@@ -240,21 +254,48 @@ class PageListViewModel @Inject constructor(
                     } else {
                         DEFAULT_INDENT
                     }
-                    PublishedPage(it.remoteId, it.title, it.date, labels, pageItemIndent,
-                            getFeaturedImageUrl(it.featuredImageId), actionsEnabled)
+                    val progressState = progressHelper.getProgressStateForPage(LocalId(it.pageId))
+
+                    PublishedPage(
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: We might want to consider reverting these types of line-breaks we added in this PR (even the method declaration). Usually, I wouldn't mind too much, but since we are merging this into a feature branch and we plan to do a big clean-up before we merge the feature branch into develop. I think it'll make it easier for us to compare which components/parts make sense to re-use between PostList and PageList. Wdyt?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt [hunk 2]
```diff
@@ -240,21 +254,48 @@ class PageListViewModel @Inject constructor(
...
-                                getFeaturedImageUrl(it.featuredImageId), actionsEnabled) }
+                .map { (date, results) ->
+                    listOf(Divider(date)) +
+                            results.map {
+                                val labels = mutableListOf<Int>()
+                                if (it.hasLocalChanges)
+                                    labels.add(R.string.local_changes)
+
+                                val progressState = progressHelper.getProgressStateForPage(
+                                        LocalId(
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: Do you have line length set to 120 characters for kotlin files?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt [hunk 3]
```diff
@@ -240,21 +254,48 @@ class PageListViewModel @Inject constructor(
...
-                            ScheduledPage(it.remoteId, it.title, it.date, labels,
-                                getFeaturedImageUrl(it.featuredImageId), actionsEnabled) }
+                .map { (date, results) ->
+                    listOf(Divider(date)) +
+                            results.map {
+                                val labels = mutableListOf<Int>()
+                                if (it.hasLocalChanges)
+                                    labels.add(R.string.local_changes)
+
+                                val progressState = progressHelper.getProgressStateForPage(
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: We might want to consider using this neat trick
`val (progressBarUiState, showOverlay)  = progressHelper.getProgressStateForPage(LocalId(it.pageId))`

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt [hunk 4]
```diff
@@ -166,26 +169,29 @@ class PagesViewModel
         val post: PostModel,
         val previewType: RemotePreviewType
     )
-
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: NitPick: We might want to revert this change.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package org.wordpress.android.viewmodel.posts
+
+sealed class PostListItemProgressBar(val visibility: Boolean) {
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt]: We might want to consider renaming this class to ProgressBarUiState and moving it into a different package. It seems it can be reused for any feature. Wdyt?

File: WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt [hunk 6]
```diff
@@ -186,6 +203,48 @@ class PageListViewModelTest : BaseUnitTest() {
         assertThat((result[0].first[1] as PublishedPage).title).isEqualTo(secondPage.title)
     }
 
+    @Test
+    fun `checks if PageItems gets the correct progress state`() {
```
Comment (malinajirka) [inline WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt]: We might want to consider breaking up this test so each test method has only a single reason to fail. We are testing multiple unrelated things within one test method. Wdyt?

Comment (jd-alexander) [reply WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt]: Okay. Thanks for this. I thought just testing the behavior of the progress state was a single thing. I will revisit 😄

Comment (jd-alexander) [reply WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt]: Oh, the test is checking the state of two different pages and behavior, right? I am going to refactor and put it into two separate tests. Thanks.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListUploadStatusTracker.kt [hunk 7]
```diff
@@ -7,12 +7,13 @@ import org.wordpress.android.fluxc.store.UploadStore
...
 import org.wordpress.android.ui.uploads.UploadService
 import org.wordpress.android.viewmodel.posts.PostListItemUploadStatus
+import javax.inject.Inject
 
 /**
  * This is a temporary class to make the PostListViewModel more manageable. Please feel free to refactor it any way
  * you see fit.
  */
-class PostListUploadStatusTracker(
+class PostListUploadStatusTracker @Inject constructor(
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListUploadStatusTracker.kt]: Good catch 😉

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemProgressHelper.kt [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }
+
+    private fun shouldShowOverlay(uploadUiState: PostUploadUiState): Boolean {
+        // show overlay when post upload is in progress or (media upload is in progress and the user is not using Aztec)
+        return (uploadUiState is UploadingPost ||
+                (!appPrefsWrapper.isAztecEditorEnabled && uploadUiState is UploadingMedia))
+    }
+
+    fun getProgressStateForPage(
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemProgressHelper.kt]: Yes, that's a good idea. It lessens the mental load and shows what's most important first. Good principle!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt [hunk 9]
```diff
@@ -694,36 +698,25 @@ class PagesViewModel
...
 
-    @Subscribe(threadMode = ThreadMode.BACKGROUND)
-    @Suppress("unused")
-    fun onEventBackgroundThread(event: PostEvents.PostUploadStarted) {
-        if (!event.post.isPage) {
-            return
+    private fun handleInvalidateUploadStatus(ids: List<LocalId>) {
+        launch {
+            _invalidateUploadStatus.postValue(ids)
+            refreshPages()
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Indeed. Good observation. So, in this case, the list just needs to be invalidated so that the upload status can be updated right?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt (hunk 5 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt]: Yes, thanks.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 3 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: Cool idea!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 1 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: Thanks. Is there any way I can modify Android Studio to avoid this because auto formatting introduces it at times. I will double-check diff for them.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 2 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: Let me check this. Thanks.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemProgressHelper.kt [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        pageId: LocalId
+    ): Pair<PostListItemProgressBar, ShouldShowOverlay> {
+        val post = postStore.getPostByLocalPostId(pageId.value)
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemProgressHelper.kt]: Thanks for this. We will need to investigate this separately from this PR. I created [an issue](https://github.com/wordpress-mobile/WordPress-Android/issues/11233) to track this!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt [hunk 11]
```diff
@@ -680,12 +686,10 @@ class PagesViewModel
     private fun hasRemoteAutoSavePreviewError() = _previewState.value != null &&
             _previewState.value == PostListRemotePreviewState.REMOTE_AUTO_SAVE_PREVIEW_ERROR
 
-    @Subscribe(threadMode = ThreadMode.MAIN)
-    @SuppressWarnings("unused")
-    fun onPostUploaded(event: OnPostUploaded) {
+    fun handlePostUploadedWithoutError(remotePostId: RemoteId) {
         var id = 0L
         if (!pageUpdateContinuations.contains(id)) {
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Honestly, no it doesn't.  I just migrated and left it because I thought it was outside the scope of the refactor but thanks for pointing it out. I am going to investigate why it was introduced or what the purpose of it is supposed to be 🙏

[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: I agree it's out-side of scope. But if the logic doesn't make sense, I'd consider creating an issue.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Yeah, I think we want to invalidate the items that were changed - just UI invalidation, there is no need to refetch their models again.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 1 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: - Preferences -> Editor -> Code Style -> HardWrap at
- Preferences -> Editor -> Code Style -> Kotlin -> Wrapping and Braces -> HardWrap at

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 2 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: Preferences -> Editor -> Code Style -> HardWrap at
Preferences -> Editor -> Code Style -> Kotlin -> Wrapping and Braces -> HardWrap at

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 1 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: Ohh I posted the reply to a wrong comment :D.

I usually just select the method I added and run the auto-formatter only on the method.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Thanks! Good idea https://github.com/wordpress-mobile/WordPress-Android/issues/11244

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListUploadStatusTracker.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListUploadStatusTracker.kt]: Resolved in e5e52fc

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemProgressHelper.kt [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+import org.wordpress.android.viewmodel.pages.PageItemProgressHelper.PostUploadUiState.UploadWaitingForConnection
+import org.wordpress.android.viewmodel.pages.PageItemProgressHelper.PostUploadUiState.UploadingMedia
+import org.wordpress.android.viewmodel.pages.PageItemProgressHelper.PostUploadUiState.UploadingPost
+import org.wordpress.android.viewmodel.posts.PostListItemProgressBar
+import org.wordpress.android.viewmodel.posts.PostListItemUploadStatus
+import javax.inject.Inject
+
+typealias ShouldShowOverlay = Boolean
+
+class PageItemProgressHelper @Inject constructor(
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemProgressHelper.kt]: Resolved in 7b8eadc

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemProgressHelper.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemProgressHelper.kt]: Resolved in 92130c0

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Resolved in 34f91c5

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 1 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: Okay, thanks!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 2 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: AWESOME!!!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 3 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: Resolved in 53cd03a

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 4 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Reverted!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt (hunk 5 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt]: Resolved in 09ddc74

File: WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt (hunk 6 above)
Comment (jd-alexander) [reply WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt]: Resolved in 175a7c3

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt [hunk 13]
```diff
@@ -227,6 +233,10 @@ class PagesViewModel
         pageMap = pageStore.getPagesFromDb(site).associateBy { it.remoteId }
     }
 
+    private fun invalidatePages() {
```
Comment (jd-alexander) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Let me know if this approach is okay to invalidate the list. I thought of it because it was a simple way and I also noticed that  a similar approach is being used to update a pages here 
https://github.com/wordpress-mobile/WordPress-Android/blob/47024f77e7ee718234bb402b05ddbe87957f85cd/WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt#L538

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt (hunk 1 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListViewModel.kt]: I will do a final run-through tomorrow again to resolve all of these. When I added the extra parameters and did the formatting they were introduced even though I ran it on the line only. I think it's because they can't hold in a single line it opted to put each on its own line.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Can you please elaborate how is this addition of `handlePostUploadedWithoutError.invoke(RemoteId((event.causeOfChange as UpdatePost).remotePostId))` related to the change? Imo `UpdatePost` event doesn't necessarily mean we've finished upload. Thanks!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt (hunk 5 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt]: All the variables now contain word "state" twice - eg. `progressBarUiStateState`. Is that intentional?

File: WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt (hunk 6 above)
Comment (malinajirka) [reply WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt]: Awesome, it's much better now. Thank you!

We might want to consider two more changes
1. Moving the most complicated test after the two simpler tests.
2. Mocking things we don't need for the test and creating local fields only from variables we actually need to use in the test

Something like this

```
 @Test
    fun `showOverlay is correctly propagated from PageItemUploadProgressHelper`() {
        // Arrange
        val expectedShowOverlay = true
        val pages = MutableLiveData<List<PageModel>>()

        whenever(progressHelper.getProgressStateForPage(any())).thenReturn(Pair(mock(), expectedShowOverlay))
        whenever(pagesViewModel.pages).thenReturn(pages)

        viewModel.start(PUBLISHED, pagesViewModel)
        val result = mutableListOf<Pair<List<PageItem>, Boolean>>()
        viewModel.pages.observeForever { result.add(it) }
        // Act
        pages.value = listOf(buildPageModel(0))
        // Assert
        assertThat((result[0].first[0] as Page).showOverlay).isEqualTo(expectedShowOverlay)
    }
```

Wdyt?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 13 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Yeah, I think it's a valid approach. It's a bit more readable in Java as it's obvious there that we are using a custom setter. But I wouldn't worry about that too much.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt [hunk 14]
```diff
@@ -90,9 +86,13 @@ class PagesViewModel
     private val eventBusWrapper: EventBusWrapper,
     private val previewStateHelper: PreviewStateHelper,
     private val uploadStarter: UploadStarter,
+    private val pageListEventListenerFactory: PageListEventListener.Factory,
     @Named(UI_THREAD) private val uiDispatcher: CoroutineDispatcher,
     @Named(BG_THREAD) private val defaultDispatcher: CoroutineDispatcher
-) : ScopedViewModel(uiDispatcher) {
+) : ScopedViewModel(uiDispatcher), LifecycleOwner {
+    private val lifecycleRegistry = LifecycleRegistry(this)
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: This creates a lifecycle, but we still need to manually change the states of the lifecycle by calling `lifecycleRegistry.markState(...)`. 

I think it might not be a best practice to implement Lifecycle in VMs - I understand you used the same approach as in PostList. It's up to you if you want to keep the `PageListEventListener` consistent with `PostListEventListener` and use LifeCycle in both or don't implement the Lifecycle and simply call `postListEventListener.onDestroy` from PagesViewModel manually.

**Comment by jd-alexander** [comment]
Thanks for the review! 

1. I was able to reproduce the crash after a few tries. I will look into it. 
2. Reproduced and looking into it. 
3. It seems that because the page is being reloaded from the server there is a delay before the statuses are updated.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemUploadProgressHelper.kt [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    private val appPrefsWrapper: AppPrefsWrapper,
+    private val postStore: PostStore,
+    val uploadStatusTracker: PostModelUploadStatusTracker
+) {
+    private lateinit var site: SiteModel
+
+    /**
+     * This was done so that this class could be injectable for testing, as this dependency wouldn't be provided by DI
+     */
+    fun attachSite(site: SiteModel) {
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemUploadProgressHelper.kt]: Good idea! I had done this because I was passing it via the constructor but I will move it to the method. Thanks.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+    fun onEventBackgroundThread(event: PostEvents.PostUploadCanceled) {
+        if (site.id == event.post.localSiteId) {
+            uploadStatusChanged(LocalId(event.post.id))
+        }
+    }
+
+    @Suppress("unused")
+    @Subscribe(threadMode = BACKGROUND)
+    fun onEventBackgroundThread(event: VideoOptimizer.ProgressEvent) {
+        uploadStatusChanged(LocalId(event.media.localPostId))
```
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt]: Thanks! Upon looking at the events again I am thinking I should also check in each event if it's a page as well since posts could be enqueued and the events could be propagated to these methods. 

`if (site.id == event.post.localSiteId && event.post.isPage) {`

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 14 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Okay! I am thinking to just call the `onDestroy` manually. Thanks for pointing this out. 🙏

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 13 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Okay! Great.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt (hunk 5 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemProgressBar.kt]: No, it wasn't. It occurred during a refactor of the name and I fixed it in a newer commit. Thanks for spotting nonetheless. You have a keen eye 😉

File: WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt (hunk 6 above)
Comment (jd-alexander) [reply WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt]: Nice! This improvement is like magic 😉

**Comment by jd-alexander** [comment]
I fixed the issues below : 
I also tested the app and I encountered following issues:

1. Open Page List
Click on the overflow menu on one of the items
Click eg. on "Move to drafts" or "Publish Now"
Boom
ebc2cd6


2. When I open a page -> change title -> click on back -> the progress indicator never disappears.
22525dd

3. Probably not related to this PR, but I'm curious if you know why. We hide the progress indicator when the upload finishes -> how come the "local changes" label disappears after another +- 3 seconds?
42faec8

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 14 above)
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: Resolved in 23c8df4

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt]: Resolved in 7fb606e

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemUploadProgressHelper.kt (hunk 15 above)
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageItemUploadProgressHelper.kt]: Resolved in 08be0df

File: WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt (hunk 6 above)
Comment (jd-alexander) [reply WordPress/src/test/java/org/wordpress/android/viewmodel/pages/PageListViewModelTest.kt]: Resolved in aecfb9f

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt]: Thanks! What about `fun onEventBackgroundThread(event: UploadService.UploadMediaRetryEvent)`. Shouldn't we also filter out mediaModels which belong to a different site? Wdyt?

[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt]: >if it's a page as well since posts could be enqueued and the events could be propagated to these methods

Good idea! We might want to create a separate PR and add `!isPage` check to `PostListEventListener`, wdyt?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt (hunk 14 above)
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PagesViewModel.kt]: 👍 Yep I agree it's better. Good call. Thanks!

**Comment by malinajirka** [comment]
Thanks for the changes @jd-alexander! I know you said you haven't had a chance to do thorough tests after the latest changes. I decided to review the PR anyway so we speed things up a bit. 

>1. ebc2cd6

Have you investigated how come the post is null? I wouldn't expect it to be null after clicking on the overflow menu. 
I'm a bit worried the null check might just hide a potential issue. The database just returned us a list of pages and a few ms later it seems the page is not in the database anymore. Do we remove the page from the database before we move it to draft/publish it?

> 2. 
👍 

>3. 42faec8

If I understand the fix correctly, we won't hide the progress indicator when the upload completes and we let the existing logic handle the refresh of the item, right? Tbh I'm getting a bit lost in the logic - it feels weird that we don't need to invalidate the status when upload completes successfully. Would you mind discussing this on a call. Thanks!



I've encountered the following crash
- I'm not able to reproduce it. I created a post, added a title and I clicked on "Up" button in the editor.
```
2020-02-07 11:45:13.904 E: FATAL EXCEPTION: DefaultDispatcher-worker-4
    Process: org.wordpress.android.beta, PID: 5736
    java.lang.IllegalStateException: Already resumed, but proposed with update kotlin.Unit
        at kotlinx.coroutines.CancellableContinuationImpl.alreadyResumedError(CancellableContinuationImpl.kt:332)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:327)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:247)
        at org.wordpress.android.viewmodel.pages.PagesViewModel.handlePostUploadedWithoutError(PagesViewModel.kt:701)
        at org.wordpress.android.viewmodel.pages.PagesViewModel$start$1.invoke(PagesViewModel.kt:185)
        at org.wordpress.android.viewmodel.pages.PagesViewModel$start$1.invoke(PagesViewModel.kt:76)
        at org.wordpress.android.viewmodel.pages.PageListEventListener$onPostChanged$1.invokeSuspend(PageListEventListener.kt:100)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:561)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:727)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:667)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:655)
```

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt]: Yes, that's a really good idea!

[the remark this answers is not in the document]
Comment (jd-alexander) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt]: Issue created here https://github.com/wordpress-mobile/WordPress-Android/issues/11259

**Comment by jd-alexander** [comment]
Thanks for the review! 
> Have you investigated how come the post is null? I wouldn't expect it to be null after clicking on the overflow menu.
I'm a bit worried the null check might just hide a potential issue. The database just returned us a list of pages and a few ms later it seems the page is not in the database anymore. Do we remove the page from the database before we move it to draft/publish it?

It doesn't seem like this is the case. Based on a glance I just did at the Flux C code, an update query is run. Nonetheless, I created an issue to track this and continue the investigation. https://github.com/wordpress-mobile/WordPress-Android/issues/11260 

> If I understand the fix correctly, we won't hide the progress indicator when the upload completes and we let the existing logic handle the refresh of the item, right? Tbh I'm getting a bit lost in the logic - it feels weird that we don't need to invalidate the status when upload completes successfully. Would you mind discussing this on a call. Thanks!

I understand this clearly. I did the invalidation but I am seeing what is causing the error now. I am going to rename my functions.

**Comment by jd-alexander** [comment]
Thanks @malinajirka I created issues for all the outstanding bugs that need to be resolved based on our exploration. I will merge this now since they are being tracked. Thanks for the reviews! Now time to add the other PRs that depend on the `UploadStatus` 🙏

File: WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (malinajirka) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageListEventListener.kt]: 👍 Thanks!
